### PR TITLE
bot, loader, plugin, rules: add NOTICE message to rate limits

### LINKED ISSUE #830: Plugin API: Add optional message to module.rate()
If set, send `message` as a notice to the calling user if the command was canceled due to rate-limit. Ideally, it would support a placeholder such as `{time_left}` for the message to tell the user when they can use the command again.

I'd be happy to code it up and PR it; just want to establish that it'd have a chance of being merged first, before putting in that time.

~~Could also have a third parameter to toggle between using NOTICE and PRIVMSG.~~

### PR DESCRIPTION
**Pull request by Exirel**
### Description

Fix #830 by adding an optional message and three new decorators for specific rate limit messages (user, channel, and server).

### Checklist

- [x] I have read [CONTRIBUTING.md](https://github.com/sopel-irc/sopel/blob/master/CONTRIBUTING.md)
- [x] I can and do license this contribution under the EFLv2
- [x] No issues are reported by `make qa` (runs `make quality` and `make test`)
- [x] I have tested the functionality of the things this change touches

### CODE REVIEW

File: sopel/bot.py
Comment (dgw) [inline sopel/bot.py:591]: As an aside, I've come to dislike `trigger.sender` as an attribute name, because it sounds like that should be the `nick` who triggered the command. While you're creating this variable and updating usage of the attribute below, can this be called `channel` instead?

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:840]:
```suggestion
    :param message: optional keyword argument; default message sent as NOTICE
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:882]:
```suggestion
        You can control each rate limit separately, with their own custom
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:907]:
```suggestion
    :param message: optional; message sent as NOTICE when a user hits the limit
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:938]:
```suggestion
    :param message: optional; message sent as NOTICE when a user hits the limit
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:950]: I have not applied this to every instance, as I'm not 100% sure the line-wrapping aligns with the standard, so I'm leaving it to you to tweak this if necessary. 😸

```suggestion
    If you don't provide a message, the default message set by :func:`rate`
    (if any) will be used instead.
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:969]:
```suggestion
    :param message: optional; message sent as NOTICE when a user hits the limit
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:965]: Not channel. This is `rate_server`.

Also worth considering if this should be called `rate_global` to match the internal terminology. References to this type of limit in `plugins.rules` also appear to call this "global".

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:978]:
```suggestion
            # as other rate limits don't have any message set
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:947]:
```suggestion
            # as other rate limits don't have any message set
```

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:916]:
```suggestion
            # as other rate limits don't have any message set
```

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:697]:
```suggestion
        :return: A formatted string, or ``None`` if no message is set.
```

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:715]:
```suggestion
        :return: A formatted string, or ``None`` if no message is set.
```

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:708]:
```suggestion
        :return: A formatted string, or ``None`` if no message is set.
```

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:807]: You know you want to, for consistency. 😏

```suggestion
            'user_rate_message': getattr(
                handler, 'user_rate_message', None),
```

File: sopel/plugins/rules.py
Comment (dgw) [inline sopel/plugins/rules.py:1121]: Since there's already some support for placeholders, what do you think of adding one for the time remaining until the rate limit expires? It's the main reason that _I_ wished for this feature to be built in, all those years ago.

Regardless of the above, these placeholders don't appear to be documented anywhere.

**Review by dgw — COMMENTED** [review]
I haven't dived into the tests yet (there are a lot of them), but there might be enough to discuss already without them.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:591]: The thing is: it could be a nick or a channel. So on one hand I agree with you, I don't like "sender", but on the other `channel` is not always accurate either.

Something like `origin` or `source`? `maybe_channel_maybe_nick`? :grin:

File: sopel/plugin.py
Comment (Exirel) [reply sopel/plugin.py:965]: Yeah, I know, I didn't change the internal terminology because I'm not sure it's the right term or not, so I've only the decorator name for now. What do you prefer? Any opinion? I can change that to `rate_global` or change internals to `server`.

(the docstring is a bad copy/paste)

File: sopel/plugins/rules.py
Comment (Exirel) [reply sopel/plugins/rules.py:1122]: Yup I thought about it. I wanted to put the bulk of the code first for review, it would be easy to add anything after that.

File: sopel/plugin.py
Comment (dgw) [reply sopel/plugin.py]: I would not at all mind calling the decorator `rate_global` instead. Somehow `rate_server` makes it sound like the server (IRCd) will be calling commands all on its own… Doesn't make sense, but that's what it sounds like!

File: sopel/bot.py
Comment (dgw) [reply sopel/bot.py:591]: `context`?

File: sopel/plugin.py
Comment (Exirel) [reply sopel/plugin.py]: Done!

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py]: Done!

File: sopel/bot.py
Comment (dgw) [inline sopel/bot.py:592]: This interface looks weird when you know that `context` already knows if it's a channel (`and not context.is_nick()`). Is it worth suggesting a change? No. But it still made me stop and think.

File: sopel/bot.py
Comment (dgw) [inline sopel/bot.py:596]: Since we've documented the Rule internals as subject to change without warning, any thoughts on taking this opportunity to rename this method as `rule.is_user_rate_limited` so it's parallel to its `get_user_rate_message` counterpart?

I'm not sure yet if there would be any utility in a replacement `rule.is_rate_limited` that could check multiple types at once, but that would be the logical choice.

File: sopel/bot.py
Comment (dgw) [inline sopel/bot.py:615]: Should this be `if is_channel and context in self.config`? I realize it's already possible to make a "Channel configuration" for some specific nick because of how this works, but the question is whether that's a bug/needs changing. Triggering this section of code in PM from someone whose nick matches a plugin's section name, for instance, feels squicky.

File: sopel/loader.py
Comment (dgw) [inline sopel/loader.py:51]: Is it time to remove the `rate` attribute from callables and replace it with `func.user_rate`, for consistency?

Do we even have to consider the actual attributes of a plugin callable as part of the API? I hope not, because this is not something I want to go through a deprecation cycle for.

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:830]: Looks weird. Not that the alternatives (`*,` at the end of the previous line, or on its own line) look great either, but let me try this on for size:

```suggestion
    *,
    message: typing.Optional[str] = None,
```

(Nope, it still looks weird. Damn it.)

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:878]: Maybe include this note on all related decorators? Even those of us who work in this codebase on a regular basis keep forgetting about it. 😁

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:919]: `rate_channel`'s format wins.

```suggestion
    If you don't provide a message, the default message set by :func:`rate`
    (if any) will be used instead.
```

File: sopel/plugins/rules.py
Comment (dgw) [reply sopel/plugins/rules.py:1122]: I also just noticed that the global and default versions don't support even a `nick` placeholder. 🤔

File: test/plugins/test_plugins_rules.py
Comment (dgw) [inline test/plugins/test_plugins_rules.py:1086]: Along with what I said before about renaming the `rate` attr on callables, this kwarg could be changed to `user_rate_limit` too.

Here I am again, dropping a line note in some mostly-unrelated place because GitHub won't let me go put it [where it belongs](https://github.com/Exirel/sopel/blob/66bae33d60c35dd77460347529bf578f00961812/sopel/plugins/rules.py#L902) on code in `sopel/plugins/rules.py` that's too far from the diff.

**Review by dgw — CHANGES_REQUESTED** [review]
Yes, I realize some of this could fall in "future PR" territory. One of us can make a tracking issue(s) once we agree on what's out of scope.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:596]: I totally agree. I'll go for the rename then.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:615]: Yeah I know, it's weird. I don't know how to feel about that. I remember that the reminder plugin didn't mind using "sender" even in PM. So I don't know yet.

File: sopel/plugin.py
Comment (Exirel) [reply sopel/plugin.py:830]: Ah ah I know right?

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:592]: After double checking, I can use context instead of trigger here.

File: sopel/bot.py
Comment (Exirel) [reply sopel/bot.py:615]: Fixed.

File: sopel/loader.py
Comment (Exirel) [reply sopel/loader.py]: And it is done.

**Comment by Exirel** [comment]
I've modified `func.rate` into `func.user_rate` (and many little changes like that).

Also now there are conflicts... I didn't look at them yet.

**Review by dgw — APPROVED** [review]
Merge conflicts are from a bit of CTCP stuff and some type-hint tweaks, AFAICT. Shouldn't be a huge deal to fix up. What's important is, this looks ready to merge so our brave `master` testers can try to break it in prod. 😁

**Comment by Exirel** [comment]
> 1) this isn't squashed/rebased

This part: done. I didn't add anything yet, because I think it would be better for you if you wanted to check the diff between what you approved so far, and what will come next (docstrings, placeholders, etc.).

**Comment by Exirel** [comment]
>  Since there's already some support for placeholders, what do you think of adding one for the time remaining until the rate limit expires? It's the main reason that I wished for this feature to be built in, all those years ago.
> 
> Regardless of the above, these placeholders don't appear to be documented anywhere.

I'm about to push a commit with just `nick` for user and global rate limits, `nick` and `channel` for channel rate limit.

For other placeholders, I need to think a bit more about it: can we assume the time left will be in seconds, and the rate limit in seconds as well? Like, if the rate limit is something like 60s/3600s, should we be smart about it? Should we provide a human readable version of that? (Using cool functions from `tools.time` for example?)

This leads me to think that, maybe, I could just commit what I have, and consider this PR "done": anyone could jump into the code and add the new placeholders as they wanted, with a separate/dedicated conversation about it, without all the other diff around.

File: sopel/plugin.py
Comment (dgw) [inline sopel/plugin.py:894]: This Unicode single quote should just be a regular old ASCII `'`.

Yes I know it was already there in the old line you cut/copied, but we do like fixing weird stuff whenever we're already touching it. There are other occurrences elsewhere in this file, so I recommend a global find-and-replace in your IDE/editor.

If you're feeling ambitious you can do a quick search through the other files this touches to find any other instances, but I wouldn't expect any.

File: test/test_bot.py
Comment (dgw) [inline test/test_bot.py:665]: Why's this deleted? It came along recently, in a2a8ba0fb65fedfc39b7b7b9d639ba90a5a740df (from #2297), so it's probably important.

**Review by dgw — COMMENTED** [review]
Looks like a small bit of collateral damage from the rebase happened; see line note.

We can certainly defer adding more kinds of placeholder for a later PR, but I really would like this patch to document the ones that it _does_ add. 🙂

File: test/test_bot.py
Comment (Exirel) [reply test/test_bot.py:665]: Oh! It's because I messed up in #2297 at some point in my development process and it wasn't necessary at all, but I forgot to remove it. It's the equivalent of ` print('debug')` but very sneaky.

**Comment by Exirel** [comment]
Squashed & rebased, as per @dgw's validation on IRC.

**Review by dgw — APPROVED** [review]
I should probably test this one last time on a real bot before merging, but that doesn't mean I can't approve it first.

**Pull request merged**
